include/vara/Feature/Feature.h

```diff
@@ -203,34 +136,90 @@ class Feature {
   }
   [[nodiscard]] const_feature_iterator end() const { return children_end(); }
 
-  [[nodiscard]] Location* getLocation() {
+  //===--------------------------------------------------------------------===//
+  // Operators
+  Feature &operator=(const Feature &) = delete;
+
+  /// Compare lowercase name assuming those are unique.
+  bool operator==(const vara::feature::Feature &F) const {
+    // TODO(s9latimm): keys in FM are case sensitive
+    return getName().lower() == F.getName().lower();
+  }
+  bool operator!=(const vara::feature::Feature &F) const {
+    return !operator==(F);
+  }
+
+  /// Compare in depth first ordering.
+  bool operator<(const vara::feature::Feature &F) const;
+  bool operator>(const vara::feature::Feature &F) const {
+    return F.operator<(*this);
+  }
+
+  [[nodiscard]] FeatureSourceRange *getFeatureSourceRange() {
     return Loc.has_value() ? &Loc.value() : nullptr;
   }
 
   //===--------------------------------------------------------------------===//
   // Utility
   [[nodiscard]] virtual std::string toString() const;
 
-  bool operator==(const Feature &F) const {
-    return llvm::StringRef(this->getName()).lower() ==
-           llvm::StringRef(F.getName()).lower();
-  }
+  void print(std::ostream &Out) const { Out << toString() << std::endl; }
+  void print(llvm::raw_ostream &Out) const { Out << toString() << '\n'; }
 
-  bool operator<(const Feature &F) const {
-    return llvm::StringRef(this->getName()).lower() <
-           llvm::StringRef(F.getName()).lower();
-  }
+  LLVM_DUMP_METHOD
+  void dump() const { llvm::outs() << toString() << "\n"; }
+
+protected:
+  Feature(FeatureType T, string Name, bool Opt,
+          std::optional<FeatureSourceRange> Loc, Feature *Parent,
+          FeatureSetType Children, FeatureSetType Excludes,
+          FeatureSetType Implications, FeatureSetType Alternatives)
+      : T(T), Name(std::move(Name)), Opt(Opt), Loc(std::move(Loc)),
+        Parent(Parent), Children(std::move(Children)),
+        Excludes(std::move(Excludes)), Implications(std::move(Implications)),
+        Alternatives(std::move(Alternatives)) {}
+
+private:
+  friend class FeatureModel;
+  friend class FeatureModelBuilder;
+
+  FeatureType T;
+  string Name;
+  bool Opt;
+  std::optional<FeatureSourceRange> Loc;
+  Feature *Parent;
+  FeatureSetType Children;
+  FeatureSetType Excludes;
+  FeatureSetType Implications;
+  FeatureSetType Alternatives;
+
+  void addChild(Feature *F) { Children.insert(F); }
+
+  void setParent(Feature *F) { Parent = F; }
+
+  void addExclude(Feature *F) { Excludes.insert(F); }
+
+  void addAlternative(Feature *F) { Alternatives.insert(F); }
+
+  void addImplication(Feature *F) { Implications.insert(F); }
```

Comment:
```suggestion
  void addChild(Feature *F) { Children.insert(F); }

  void setParent(Feature *F) { Parent = F; }

  void addExclude(Feature *F) { Excludes.insert(F); }

  void addAlternative(Feature *F) { Alternatives.insert(F); }

  void addImplication(Feature *F) { Implications.insert(F); }
  
  FeatureType T;
  string Name;
  bool Opt;
  std::optional<FeatureSourceRange> Loc;
  Feature *Parent;
  FeatureSetType Children;
  FeatureSetType Excludes;
  FeatureSetType Implications;
  FeatureSetType Alternatives;
```
Nit: I would prefer to have all members at the end of the class.